assumeutxo: Get rid of faked nTx and nChainTx values

ryanofsky opened the pull request:
The `PopulateAndValidateSnapshot` function introduced in f6e2da5fb7c6406c37612c838c998078ea8d2252 from #19806 has been setting fake `nTx` and `nChainTx` values that can show up in RPC results (https://github.com/bitcoin/bitcoin/issues/29328) and make `CBlockIndex` state hard to reason about, because it is difficult to know whether the values are real or fake.

Revert to previous behavior of setting `nTx` and `nChainTx` to 0 when the values are unknown, instead of faking them. Also drop no-longer needed `BLOCK_ASSUMED_VALID` flag.

Dropping the faked values also fixes assert failures in the `CheckBlockIndex` `(pindex->nChainTx == pindex->nTx + prev_chain_tx)` check that could happen previously if forked or out-of-order blocks before the snapshot got submitted while the snapshot was being validated. The PR includes two commits adding tests for these failures and describing them in detail.

Compatibility note: This change could cause new `-checkblockindex` failures if a snapshot was loaded by a previous version of Bitcoin Core and not fully validated, because fake `nTx` values will have been saved to the block index. It would be pretty easy to avoid these failures by adding some compatibility code to `LoadBlockIndex` and changing `nTx` values from 1 to 0 when they are fake (when `(pindex->nStatus & BLOCK_VALID_MASK) < BLOCK_VALID_TRANSACTIONS`), but a little simpler not to worry about being compatible in this case.

ryanofsky committed:
test: assumeutxo stale block CheckBlockIndex crash test

Add a test for a CheckBlockIndex crash that would happen before previous
"assumeutxo: Get rid of faked nTx and nChainTx values" commit.

The crash was an assert failure in the (pindex->nChainTx == pindex->nTx +
prev_chain_tx) check that would previously happen if a snapshot was loaded, and
a block was submitted which forked from the chain before the snapshot block and
after the last downloaded background chain block. This block would not be
marked assumed-valid because it would not be an ancestor of the snapshot, and
it would have nTx set, nChainTx unset, and prev->nChainTx set with a fake
value, so the assert would fail. After the fix, prev->nChainTx is unset instead
of being set to a fake value, so the assert succeeds. This test was originally
posted by maflcko in
https://github.com/bitcoin/bitcoin/issues/29261#issuecomment-1918947945

Co-authored-by: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>

ryanofsky committed:
doc: Improve comments describing setBlockIndexCandidates checks

The checks are changing slightly in the next commit, so try to explains the
ones that exist to avoid confusion
(https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499519079)

ryanofsky committed:
test: assumeutxo snapshot block CheckBlockIndex crash test

Add a test for a CheckBlockIndex crash that would happen before previous
"assumeutxo: Get rid of faked nTx and nChainTx values" commit.

The crash was an assert failure in the (pindex->nChainTx == pindex->nTx +
prev_chain_tx) check that would previously happen if the snapshot block was
submitted after loading the snapshot and downloading a few blocks after the
snapshot. In that case ReceivedBlockTransactions() previously would overwrite
the nChainTx value of the submitted snapshot block with a fake value based on
the previous block, so the (pindex->nChainTx == pindex->nTx + prev_chain_tx)
check would later fail on the first block after the snapshot. This test was
originally posted by Martin Zumsande <mzumsande@gmail.com> in
https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

ryanofsky committed:
assumeutxo: Get rid of faked nTx and nChainTx values

The `PopulateAndValidateSnapshot` function introduced in
f6e2da5fb7c6406c37612c838c998078ea8d2252 from #19806 has been setting fake
`nTx` and `nChainTx` values that can show up in RPC results (see #29328) and
make `CBlockIndex` state hard to reason about, because it is difficult to know
whether the values are real or fake.

Revert to previous behavior of setting `nTx` and `nChainTx` to 0 when the
values are unknown, instead of faking them.

This commit fixes at least two assert failures in the (pindex->nChainTx ==
pindex->nTx + prev_chain_tx) check that would happen previously. Tests for
these failures are added separately in the next two commits.

Compatibility note: This change could result in -checkblockindex failures if a
snapshot was loaded by a previous version of Bitcoin Core and not fully
validated, because fake nTx values will have been saved to the block index. It
would be pretty easy to avoid these failures by adding some compatibility code
to `LoadBlockIndex` and changing `nTx` values from 1 to 0 when they are fake
(when `(pindex->nStatus & BLOCK_VALID_MASK) < BLOCK_VALID_TRANSACTIONS`), but a
little simpler not to worry about being compatible in this case.

ryanofsky commented:
> Not sure. This can only happen on test-chains, so I think it would be cleaner to just require the developers, if there are any, to finish their background sync, or to wipe their chainstate, or the blocksdir, or the whole test data dir.

It makes sense that having backwards compatibility may not be worth it, but I don't see why test chains would set `nTx` to a real value without also setting BLOCK_VALID_TRANSACTIONS. It seems like we should be able to use BLOCK_VALID_TRANSACTIONS to know if nTx is real or fake, as long as we are checking for BLOCK_VALID_TRANSACTIONS, directly, and not using the IsValid() helper which is also influenced by BLOCK_FAILED flags.

---

Updated 4c70d993e3ac0e9b7923eded51826fe1faa2d004 -> a3228f02f69e28960f7da76054965c916e781f27 ([`pr/nofake.1`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.1) -> [`pr/nofake.2`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.1..pr/nofake.2)) to fix broken validation_block_tests due to a buggy check, also making some minor cleanups

maflcko reviewed (approved):
lgtm

ryanofsky commented on src/validation.cpp:
```diff
@@ -4903,36 +4890,23 @@ void ChainstateManager::CheckBlockIndex()
-            // Assumed-valid blocks should have some nTx value.
-            assert(pindex->nTx > 0);
-            // Assumed-valid blocks should connect to the main chain.
-            assert((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TREE);
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1478022794

> Any reason to remove this check?

No, good catch, I'll add it back. I was thinking it wasn't really related to the other checks here, and trying to eliminate unnecessary uses of BLOCK_ASSUMED_VALID / IsAssumedValid in general. But as long as we have the assumed valid flag, might as well check this.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4903,36 +4890,23 @@ void ChainstateManager::CheckBlockIndex()
-        assert((pindexFirstNeverProcessed == nullptr) == pindex->HaveNumChainTxs());
-        assert((pindexFirstNotTransactionsValid == nullptr) == pindex->HaveNumChainTxs());
+        // HaveNumChainTxs will also be set in the assumeutxo snapshot block from snapshot metadata.
+        const bool snap_block = pindex == GetSnapshotBaseBlock();
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1478018977

> Can this be further limited to only mark the snap_block if it has ASSUMED_VALID set? Otherwise, nodes which never used loadtxoutset will assume that some block is a "snapshot block", even though they never used that feature.

I'm pretty sure I can do that, but it should not change anything. If loadtxoutset is not used, GetSnapshotBaseBlock will return null and snap_block should already be false. The only case where the suggestion would change the checks performed (when GetSnapshotBaseBlock() == pindex && !pindex->IsAssumedValid()) is the case where background chain has been downloaded and fully validated, and the snapshot block loses it's assumed_valid flag after it become validated. In that case, none of the snap_block conditions will ever be reached. Whether a snapshot block is still considered a snapshot block after the snapshot is validated is just a semantic question.

ryanofsky reviewed (commented):
Thanks for the close review! Will work on your suggestions

ryanofsky committed:
ci: add getchaintxstats ubsan suppressions

Add ubsan suppressions for integer overflows in the getchaintxstats RPC.

getchainstatstx line "int nTxDiff = pindex->nChainTx - past_block.nChainTx" can
trigger ubsan integer overflows when assumeutxo snapshots are loaded, from
subtracting unsigned values and assigning the result to a signed int.

The overflow behavior probably exists in current code but is hard to trigger
because it would require calling getchainstatstx at the right time with
specific parameters as background blocks are being downloaded. But the overflow
behavior becomes easier to trigger in the upcoming commit removing fake
nChainTx values, so a suppression needs to be added before then for CI to pass.

getchainstatstx should probably be improved separately in another PR to not
need this suppression, and handle edge cases and missing nChainTx values more
carefully.

ryanofsky committed:
assumeutxo test: Add RPC test for fake nTx and nChainTx values

The fake values will be removed in an upcoming commit, so it is useful to have
test coverage confirming the change in behavior.

ryanofsky committed:
assumeutxo: Remove BLOCK_ASSUMED_VALID flag

Flag adds complexity and is not currently used for anything.

ryanofsky reviewed (commented):
Updated a3228f02f69e28960f7da76054965c916e781f27 -> 3bdbc3fc7f116856acc7d955f591110b96f5ebd2 ([`pr/nofake.2`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.2) -> [`pr/nofake.3`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.2..pr/nofake.3)) implementing all suggestions, adding test coverage for the RPCs, and adding a new commit that drops the no longer needed BLOCK_ASSUMED_VALID flag.
Updated 3bdbc3fc7f116856acc7d955f591110b96f5ebd2 -> 603a92c9881ed78e29f15c48121ce8bf35d30837 ([`pr/nofake.3`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.3) -> [`pr/nofake.4`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.3..pr/nofake.4)) removing more BLOCK_ASSUMED_VALID references

mzumsande commented on src/validation.cpp:
```diff
@@ -5531,14 +5536,6 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
-        if (!index->nTx) {
-            index->nTx = 1;
-        }
-        // Fake nChainTx so that GuessVerificationProgress reports accurately
```
Would `GuessVerificationProgress` lose accuracy with the removal or was the comment wrong?

mzumsande replied:
> do you have a code path where nChainTx is read when it was wrong (previously) or is unset (now)

no, I never understood how the fake `nChainTx` values would help with guessing progress, but also wasn't 100% sure.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1479169504

> unrelated to this pull: Maybe an `Assume(nchaintx)` can be added to `GuessVerificationProgress`?

I went ahead and added this in a separate commit

ryanofsky commented:
> Needs rebase

Yes, thanks for recognizing the CI failures.

Rebased 603a92c9881ed78e29f15c48121ce8bf35d30837 -> 3f0b8607f549b27e9bb0cc8b189252a5cd954a36 ([`pr/nofake.4`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.4) -> [`pr/nofake.5`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.4-rebase..pr/nofake.5)) with test fixes needed after silent conflict with #29354

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1931513793

> Could include the test [#29370 (comment)](https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1927200215) , or did you want to leave that for later?

I had just forgotten about this. Added the test now, and fixed another bug uncovered by the test (forgetting to reset pindexFirstNeverProcessed after moving upwards from the snapshot block).

Updated 3f0b8607f549b27e9bb0cc8b189252a5cd954a36 -> 2f4e7e8dc82adb60a339211cdc15408b5a6919ce ([`pr/nofake.5`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.5) -> [`pr/nofake.6`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.5..pr/nofake.6)) adding maflcko's test

ryanofsky committed:
validation: Check GuessVerificationProgress is not called with disconnected block

Use Assume macro as suggested https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1479427801

ryanofsky reviewed (commented):
Updated 2f4e7e8dc82adb60a339211cdc15408b5a6919ce -> c9ca4ca4d970f9d42d0254775c04d114bd2152c3 ([`pr/nofake.6`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.6) -> [`pr/nofake.7`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.6..pr/nofake.7)) added suggested unset nChainTx check in GuessVerificationProgress

mzumsande commented on src/chain.h:
```diff
@@ -128,21 +128,8 @@ enum BlockStatus : uint32_t {
-     * should not be used elsewhere.
-     */
-    BLOCK_ASSUMED_VALID      =   256,
+    BLOCK_STATUS_RESERVED    =   256, //!< Unused flag that was previously set on assumeutxo snapshot blocks and their
```
Is it necessary to keep a placeholder / would just removing it break something? I'm asking beacuse as far as I can see, the `enum BlockStatus` is only used in `IsValid` / `RaiseValidity`, but not for `nStatus`, which is a `uint32_t`.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1484705186

> Is it necessary to keep a placeholder / would just removing it break something? I'm asking beacuse as far as I can see, the `enum BlockStatus` is only used in `IsValid` / `RaiseValidity`, but not for `nStatus`, which is a `uint32_t`.

If this asking whether dropping the placeholder would change the underlying enum type, I think it wouldn't because it it's declared 32-bit unsigned explicitly (`enum BlockStatus : uint32_t`)

Sjors replied:
If we keep the comment, perhaps add that this has never been set for any released or master branch mainnet node. So it can be repurposed, especially if at some point in the future both signet and testnet have been reset.

mzumsande commented on doc/design/assumeutxo.md:
```diff
@@ -51,12 +51,6 @@ The utility script
 
 ## Design notes
 
-- A new block index `nStatus` flag is introduced, `BLOCK_ASSUMED_VALID`, to mark block
```
The next paragraph refers to this deleted comment ("using
  the nStatus flag mentioned above") and should be updated too.

mzumsande reviewed (commented):
Concept ACK
This looks like a nice simplification.

ryanofsky commented on test/functional/feature_assumeutxo.py:
```diff
@@ -205,6 +208,35 @@ def run_test(self):
+        assert_equal(n1.getblockheader(snapshot_prev_hash)["nTx"], 1)
+        assert_equal(n1.getblockheader(snapshot_hash)["nTx"], 1)
+
+        # After loading the snapshot, check nChainTx values indirectly, by using the
```
> Why check these values indirectly instead of directly (field `txcount`)?

Wow, I did not notice that field 🤦. That will make the test a lot better.

ryanofsky reviewed (commented):
Thanks for the review!

Updated c9ca4ca4d970f9d42d0254775c04d114bd2152c3 -> 8ab4e07a1a54c9f4818a557fd1182bbb59e1cf9b ([`pr/nofake.7`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.7) -> [`pr/nofake.8`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.7..pr/nofake.8)) with suggestions

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -216,20 +224,19 @@ def run_test(self):
         assert_equal(n1.getchaintxstats(nblocks=1, blockhash=start_hash)["txcount"], START_HEIGHT + 1)
-        assert_equal(n1.getchaintxstats(nblocks=1, blockhash=start_next_hash)["txcount"], START_HEIGHT + 2)
-        assert_equal(n1.getchaintxstats(nblocks=1, blockhash=snapshot_prev_hash)["txcount"], SNAPSHOT_BASE_HEIGHT)
+        assert_equal(n1.getchaintxstats(nblocks=1, blockhash=start_next_hash)["txcount"], 0)
```
CI failed on this line: https://cirrus-ci.com/task/5549867297144832?logs=ci#L2880

fjahr commented:
Concept ACK, looks good at first sight, will do a deeper review when it's clear if the CI failure is an issue. I couldn't reproduce it locally so far...

ryanofsky commented:
CI failure is just a preexisting bug in a line of code that doesn't affect anything:

https://github.com/bitcoin/bitcoin/blob/45b2a91897ca8f4a3e0c1adcfb30cf570971da4e/src/rpc/blockchain.cpp#L1658

Two `nChainTx` values in the `getchaintxstats` are subtracted and this triggers the undefined behavior sanitizer because they are unsigned and the result is negative. `getchaintxstats` has a number of edge case problems like this that should be cleaned up but don't really affect anything. I'll probably just add a ubsan suppression in this PR.

ryanofsky commented:
> To reproduce locally, you'll have to compile with the `integer` sanitizer (not to be confused by the `undefined` sanitizer)

Thanks! I was just trying to reproduce this locally using the undefined sanitizer.

ryanofsky commented:
Updated 8ab4e07a1a54c9f4818a557fd1182bbb59e1cf9b -> 8afcd99435b693f69b8c3a918b0573ef12559b22 ([`pr/nofake.8`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.8) -> [`pr/nofake.9`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.8..pr/nofake.9)) adding suppression to avoid getchaintxstats ubsan integer overflow errors exposed by new test coverage

maflcko reviewed (commented):
ACK 8afcd99435b693f69b8c3a918b0573ef12559b22 🤗

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 8afcd99435b693f69b8c3a918b0573ef12559b22 🤗
pDg2BPKGH9xuFZAcu+QXuXdu2fcqOaeS2y/Crj2fW2R6q67lpuPvQS+Iq51IomT4ZBoP+oCfeUSQY6QiyCRLCg==
```

</details>

ryanofsky commented on src/validation.cpp:
```diff
@@ -4944,30 +4948,33 @@ void ChainstateManager::CheckBlockIndex()
+            const bool is_active = c == &ActiveChainstate();
             if (c->m_chain.Tip() == nullptr) continue;
-            if (!CBlockIndexWorkComparator()(pindex, c->m_chain.Tip()) && pindexFirstNeverProcessed == nullptr) {
+            if (!CBlockIndexWorkComparator()(pindex, c->m_chain.Tip()) && (pindexFirstNeverProcessed == nullptr || (is_active && pindex == snap_base))) {
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499266054

> Just to clarify, the `is_active` check is just a belt-and-suspender, since `pindex == snap_base` implies `is_active`?

It's meant to be a real check. I should probably add a comment about it. The idea is that on the snapshot chain, the snapshot block should be in `setBlockIndexCandidates` even though earlier transactions may not be downloaded.

I don't think `pindex == snap_base` implies anything about `is_active`, because the `for ... GetAll` loop above should be iterating over all chains and only one of them should be active.

However, I think it might be true right now that the snapshot block is added to `setBlockIndexCandidates` for all chains when it is missing transaction data, even though it only makes sense to add it on the snapshot chain. So the checks below may happen to succeed even if `is_active` is false. But if that's the case, it's just a side-effect of the way some sloppy code around TryAddBlockIndexCandidate is written, not something we should rely on here:

https://github.com/bitcoin/bitcoin/blob/1ac627c485a43e50a9a49baddce186ee3ad4daad/src/validation.cpp#L4602-L4613

https://github.com/bitcoin/bitcoin/blob/1ac627c485a43e50a9a49baddce186ee3ad4daad/src/test/validation_chainstatemanager_tests.cpp#L519-L522

I think we will probably want a separate PR to clean up code around `TryAddBlockIndexCandidate` and only add blocks to `setBlockIndexCandidates` that really belong there.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499266054

> Just to clarify, the `is_active` check is just a belt-and-suspender, since `pindex == snap_base` implies `is_active`?
> 
> An alternative may be to split the assumption out into a separate assert.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4944,30 +4948,33 @@ void ChainstateManager::CheckBlockIndex()
                         // as a candidate, but a background chainstate should
                         // only have it if it is an ancestor of the snapshot base.
-                        if (is_active || GetSnapshotBaseBlock()->GetAncestor(pindex->nHeight) == pindex) {
+                        if (is_active || snap_base->GetAncestor(pindex->nHeight) == pindex) {
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499519079

> Why would the _background_ chainstate have the _snapshot_ base? The comment seems wrong and the second branch of the if condition is never taken, no?

I think the comment is just saying that if pindex is an ancestor of the snapshot base, it should be in setBlockIndexCandidates. It's possible the snapshot base could be in there too, but the code is only checking for it to be there if the chain is active, or if all transactions in and before the snapshot base were downloaded.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499519079

> re: [#29370 (comment)](https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499519079)
> 
> > Why would the _background_ chainstate have the _snapshot_ base? The comment seems wrong and the second branch of the if condition is never taken, no?
> 
> I think the comment is just saying that if pindex is an ancestor of the snapshot base, it should be in setBlockIndexCandidates. It's possible the snapshot base could be in there too, but the code is only checking for it to be there if the chain is active, or if all transactions in and before the snapshot base were downloaded.

ryanofsky reviewed (commented):
Thanks for the review. Tried to answer some questions below and I will work on the suggestions

ryanofsky commented on src/validation.cpp:
```diff
@@ -5115,6 +5115,11 @@ double GuessVerificationProgress(const ChainTxData& data, const CBlockIndex *pin
         return 0.0;
 
+    if (!Assume(pindex->nChainTx > 0)) {
+        LogPrintf("Internal bug detected: block %d has unset nChainTx (%s %s). Please report this issue here: %s\n",
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499040796

> second commit: LogPrintf is deprecated. Also, could use `STR_INTERNAL_BUG` to reduce boilerplate.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4944,30 +4948,33 @@ void ChainstateManager::CheckBlockIndex()
+
         // Chainstate-specific checks on setBlockIndexCandidates
         for (auto c : GetAll()) {
+            const bool is_active = c == &ActiveChainstate();
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1499230459

> nit in [594336a](https://github.com/bitcoin/bitcoin/commit/594336ae8aad026dacb5d536df63bd374e3a89c7): Would be good to avoid the use of `=` in C++, when possible, as it narrowing and possibly confusing when used in contexts with `==`.
> 
> ```c++
> const bool is_active{c == &ActiveChainstate()};
> ```

ryanofsky commented:
Updated 8afcd99435b693f69b8c3a918b0573ef12559b22 -> 7081caca03b8f0d1c96415e0fc839bf4836257be ([`pr/nofake.9`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.9) -> [`pr/nofake.10`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.9..pr/nofake.10)) adding check for final nTx and nChainTx values to the python test, moving ubsan suppressions to another commit, and improving setBlockIndexCandidates checks and adding a better comments.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4902,36 +4916,26 @@ void ChainstateManager::CheckBlockIndex()
-        assert((pindexFirstNeverProcessed == nullptr) == pindex->HaveNumChainTxs());
-        assert((pindexFirstNotTransactionsValid == nullptr) == pindex->HaveNumChainTxs());
+        // HaveNumChainTxs will also be set in the assumeutxo snapshot block from snapshot metadata.
+        assert((pindexFirstNeverProcessed == nullptr || pindex == snap_base) == pindex->HaveNumChainTxs());
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505114232

In commit "assumeutxo: Get rid of faked nTx and nChainTx values" (d351ea2f887dc81993799010ac2fb6a220f07561)

> we'll set `nChainTx` (which was previously set to the chainparams value) back to 0

Wow, great catch. It never occurred to me that we would set the `nChainTx` back to 0 after setting it to another value. But that's exactly what we do at the beginning of ReceivedBlockTransactions():

https://github.com/bitcoin/bitcoin/blob/ba907f96ad37c09c49c0e1532fad118fcb8dd4a8/src/validation.cpp#L3616

It looks like this bug was actually present in the previous assert too, and not new in this commit?

In light of that, I guess the check we would need here is:

```c++
if (pindex->HaveNumChainTxs()) {
    assert(pindexFirstNeverProcessed == nullptr || pindex == snap_base);
} else {
    assert(pindexFirstNeverProcessed != nullptr);   
}
```

But I think a better change would be to make ReceivedBlockTransactions() not reset nChainTx to 0 at all. Setting the snapshot block `nChainTx` to a real value, then back to 0, then back to a real value again seems like confusing behavior that could break other code aside from this assert.

mzumsande replied:
> It looks like this bug was actually present in the previous assert too, and not new in this commit?

Yes, I encountered this issue testing this branch on signet (and hadn't tried that with master), but it should also fail on master. So I think it would also be ok to fix this not here but in a separate PR.

Slightly related: I wonder if we should do something differently in the situation where the chainparams-provided `nChainTx` turns out to be wrong. Of course that's trusted data and shouldn't be incorrect normally - but if it is anyway I think that we'd currently continue with the inconsistent `nChainTx` data and wouldn't even log an error - unless `-checkblockindex` is enabled, in which case we'd terminate with an assert.

mzumsande replied:
> Agree it should probably be a separate PR. 

ok - I can work on a follow-up that fixes the bug (preferably in `ReceivedBlockTransactions` instead of just adjusting the checks) , and add a test case that would trigger the assert on master.

ryanofsky replied:
> ok - I can work on a follow-up that fixes the bug (preferably in `ReceivedBlockTransactions` instead of just adjusting the checks) , and add a test case that would trigger the assert on master.

That would be great, and I'd be happy to review. Let me know if you think I should update the CheckBlockIndex assert in this PR as described above https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505146598 too here.

I did experiment a little with adding asserts to ReceivedBlockTransactions:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3589,6 +3589,8 @@ void Chainstate::TryAddBlockIndexCandidate(CBlockIndex* pindex)
 void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockIndex* pindexNew, const FlatFilePos& pos)
 {
     AssertLockHeld(cs_main);
+    assert(pindexNew->nTx == 0);
+    assert(pindexNew->nChainTx == 0);
     pindexNew->nTx = block.vtx.size();
     pindexNew->nChainTx = 0;
     pindexNew->nFile = pos.nFile;
```

This makes the two assumeutxo tests fail as expected, but it also makes the `rpc_getblockfrompeer.py` test fail the `nTx == 0` check for some reason. Hopefully a straightforward fix ReceivedBlockTransactions should be possible, though.

mzumsande replied:
> Let me know if you think I should update the CheckBlockIndex assert in this PR as described above https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505146598 too here.

Up to you, but I think it'd be good to have this fixed and would be happy to re-ACK.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1508080078

> Up to you, but I think it'd be good to have this fixed and would be happy to re-ACK.

Ok, I updated the assertion in a new commit 438ac1ffc23ec5134f2ac74cee75ff145078df09, which could also be a standalone PR. (I'm assuming by "fixed" you mean changing the assertion, not changing ReceivedBlockTransactions, which I think would be harder to do correctly.)

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1508212488

> Ok, I updated the assertion in a new commit [438ac1f](https://github.com/bitcoin/bitcoin/commit/438ac1ffc23ec5134f2ac74cee75ff145078df09), which could also be a standalone PR. (I'm assuming by "fixed" you mean changing the assertion, not changing ReceivedBlockTransactions, which I think would be harder to do correctly.)

New asserts had some problems reported https://github.com/bitcoin/bitcoin/pull/29370#pullrequestreview-1912039200 with a test case triggering failures posted https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. It was possible to fix the asserts in CheckBlockIndex to cope with ReceivedBlockTransactions setting 0 nChainTx values, but the checks started getting more complicated and it turned out to be simpler to just update ReceivedBlockTransactions not to mess up nChainTx instead.

Also debugging the test case revealed that the new crash it was triggering with this PR was not the same as the crash it was triggering before this PR. So commit 438ac1ffc23ec5134f2ac74cee75ff145078df09 would not actually fix anything. The test case works by submitting a snapshot block after loading the snapshot, out of order before previous blocks were downloaded. Before this PR, this would cause ReceivedBlockTransactions to set a fake nChainTx value on the snapshot block and the `pindex->nChainTx == pindex->nTx + prev_chain_tx` check on the next block to fail. After this PR it would cause ReceivedBlockTransactions to set a 0 nChainTx value and the `assert((pindexFirstNeverProcessed == nullptr || pindex == snap_base) == pindex->HaveNumChainTxs());` check to fail.

mzumsande commented:
ACK 2f3a692cdad60cb771e390bb79bf4da2b40146c9 - I reviewed the code, and did some testing running with `-checkblockindex` on signet.

Would be great if @jamesob you could have a quick look. Not necessarily an in-depth review - I'd just want too make sure we're not missing some historical context from the time when the faked values and the `BLOCK_ASSUMED_VALID` flag were introduced originally.

ryanofsky reviewed (commented):
Updated 2f3a692cdad60cb771e390bb79bf4da2b40146c9 -> 7ce1a0e375a2b27614be0ff148aea989d1100ff8 ([`pr/nofake.11`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.11) -> [`pr/nofake.12`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.11..pr/nofake.12)) adding new commit 438ac1ffc23ec5134f2ac74cee75ff145078df09 to fix a pre-existing bug https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505114232 in CheckBlockIndex

mzumsande reviewed (commented):
The situation with a snapshot block that is sent out of order is still asserting (in two places). As mentioned before, I'd understand if you wanted to delay the fix to a different PR.

ryanofsky commented on src/validation.cpp:
```diff
@@ -4930,8 +4931,22 @@ void ChainstateManager::CheckBlockIndex()
+            assert(pindexFirstNotTransactionsValid == nullptr);
+        } else {
+            const bool unprocessed_snapshot_block{pindex == snap_base && pindex->nTx == 0};
+            assert(pindexFirstNotTransactionsValid != nullptr || unprocessed_snapshot_block);
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509552384

> I think it must be `pindex->nTx != 0` in the line above - since the situation is that we have received the snapshot block transactions (so `nTx!=0`), but can't connect the block yet (so `nChainTx==0`). Also, maybe `unconnected_snapshot_block` is a better name?

Oh, I forgot that nTx would still be set in that case. I was trying to make the check stricter so snap_base would have to have nChainTx set after it was fully connected, and it wouldn't be able to escape this check after that happened. But I guess there is no way to determine that and the best that can be done is to check that it nChainTx is nonzero after it is validated. Can make that update

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509552384

Could reproduce this failure with the test change suggested in https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. Changing this to `const bool unprocessed_snapshot_block{pindex == snap_base && (pindex->nStatus & BLOCK_VALID_MASK) < BLOCK_VALID_SCRIPTS}` did fix this but then there was just another assert failure in the `assert((pindex->nChainTx != 0) == (pindex == snap_base));` line below, so opted to revert this and make a fix in `ReceivedBlockTransactions` instead

ryanofsky commented on src/validation.cpp:
```diff
@@ -4959,14 +4965,17 @@ void ChainstateManager::CheckBlockIndex()
+            assert(pindex->nChainTx == pindex->nTx + pindex->pprev->nChainTx);
+        } else {
+            // Otherwise nChainTx should only be set if this is a snapshot block.
+            assert((pindex->nChainTx != 0) == (pindex == snap_base));
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509611419

> This is also violated in the situation of a snapshot block that is sent out of order (`pindex->nChainTx == 0`, `pindex == snap_base` )

Ok, I guess it needs to be changed to:

```c++
assert(pindex->nChainTx != 0 || pindex == snap_base);
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509611419

Could reproduce this failure with the test change suggested in https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. It actually needed a slightly different fix `assert(pindex->nChainTx == 0 || pindex == snap_base || pindex->pprev == snap_base)` because there if the snapshot block temporarily had nChainTx set to 0, and the block after the snapshot block was previously downloaded, the assert could trigger again without the `pindex->pprev == snap_base` condition. Explaining this in another code comment seemed too complicated though, so opted to revert this and make a fix in `ReceivedBlockTransactions` instead

ryanofsky reviewed (commented):
Thanks for the review. It seems like my workaround for the (preexisting) bug reported in the https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1508080078 thread doesn't work, and I'm undecided whether it's better to make tweaks to fix it or just leave it alone to be solved separately. @mzumsande I'm really curious how you're testing it, because it is a little hard to reason about abstractly without a test.

mzumsande commented:
> @mzumsande I'm really curious how you're testing it, because it is a little hard to reason about abstractly without a test.

I've been testing this in two ways:
1) On signet, syncing from a snapshot I created until I am close to the point where the background chainstate is finished, and then running with `-checkblockindex` for the last couple of blocks until it crashes. After several tries I've now saved a copy of a datadir where the background sync is finished up to `snapshot_height - 150`  or so, and then I run only the last few blocks with the slow `-checkblockindex`. The signet testing is also the source how I encountered #29519.

2) Adjusting `feature_assumeutxo.py`, so far I've just added the code 
```
        # Send snapshot block to n1 out of order, to test checkblockindex behavior
        snapshot_hash = n0.getblockhash(SNAPSHOT_BASE_HEIGHT)
        snapshot_block = n0.getblock(snapshot_hash, 0)
        n1.submitblock(snapshot_block)
```
before `self.connect_nodes(0, 1)` in line ~308. This triggers the bug at the snapshot height, but is not ideal otherwise, because it means that the order of blocks is really strange, e.g. we receive the block snapshot+1 after snapshot which would not usually happening when syncing. I've run into additional asserts failures this way (not mentioned above) that will hopefully be no longer an issue after changing the logic in `ReceivedBlockTransactions`. But I'm thinking of generalising this approach - maybe even randomize the order in which all blocks are sent - but I haven't done that yet.

ryanofsky commented:
Thank you that is very helpful! I should have thought of submitting the snapshot block to test this. For some reason I was thinking the bug would be harder to trigger.

mzumsande reviewed (commented):
Fixing the issue in ReceivedBlockTransactions turned out to be really simple, see 4cba9f4f0693ecae23f8c02afe354c5893c2684c in  [202403_assumeutxo_nchaintx](https://github.com/mzumsande/bitcoin/tree/202403_assumeutxo_nchaintx) that builds on this branch, and also adds a test to submit blocks in a randomized order.
So I'm not sure if it's really worth the effort to get the checks right in the current world where nChainTx for the snapshot block can be set back to 0.

ryanofsky commented:
> Fixing the issue in ReceivedBlockTransactions turned out to be really simple, see [4cba9f4](https://github.com/bitcoin/bitcoin/commit/4cba9f4f0693ecae23f8c02afe354c5893c2684c) in [202403_assumeutxo_nchaintx](https://github.com/mzumsande/bitcoin/tree/202403_assumeutxo_nchaintx) that builds on this branch, and also adds a test to submit blocks in a randomized order. So I'm not sure if it's really worth the effort to get the checks right in the current world where nChainTx for the snapshot block can be set back to 0.

Yeah after putting in more effort, I gave up on the checks and came to a similar conclusion. My fix is more strict, though and looks like:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3590,7 +3590,14 @@ void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockInd
 {
     AssertLockHeld(cs_main);
     pindexNew->nTx = block.vtx.size();
-    pindexNew->nChainTx = 0;
+    // Typically nChainTx will be 0 at this point, but it can be nonzero if this
+    // is a pruned block which is being downloaded again, or if this is an
+    // assumeutxo snapshot block which has a hardcoded nChainTx value from the
+    // snapshot metadata. If the pindex is not the snapshot block and the nChainTx
+    // value is not zero, assert that value is actually correct.
+    assert(pindexNew->nChainTx == 0 ||
+           pindexNew->nChainTx == (pindexNew->pprev ? pindexNew->pprev->nChainTx : 0) + pindexNew->nTx ||
+           pindexNew == GetSnapshotBaseBlock());
     pindexNew->nFile = pos.nFile;
     pindexNew->nDataPos = pos.nPos;
     pindexNew->nUndoPos = 0;
@@ -3610,7 +3617,13 @@ void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockInd
         while (!queue.empty()) {
             CBlockIndex *pindex = queue.front();
             queue.pop_front();
-            pindex->nChainTx = (pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx;
+            auto chain_tx{(pindex->pprev ? pindex->pprev->nChainTx : 0) + pindex->nTx};
+            // Before setting nChainTx assert that it is 0 or already set to
+            // the correct value. This assert will fail after receiving the
+            // assumeutxo snapshot block if assumeutxo snapshot metadata has an
+            // incorrect AssumeutxoData::nChainTx value.
+            assert(pindex->nChainTx == 0 || pindex->nChainTx == chain_tx);
+            pindex->nChainTx = chain_tx;
             pindex->nSequenceId = nBlockSequenceId++;
             for (Chainstate *c : GetAll()) {
                 c->TryAddBlockIndexCandidate(pindex);
```

mzumsande commented:
> My fix is more strict, though and looks like: (...)

Seems fine to me, though I'm not sure in favor of adding nontrivial asserts like this outside of `CheckBlockIndex` - I'd prefer them to be `Assume` in production code, since the consequences of them breaking at some point could be very serious, whereas the consequence of a wrong (non-zero) `nChainTx` value would be mostly just that progress estimation logging would be broken.

ryanofsky reviewed (commented):
Updated 7ce1a0e375a2b27614be0ff148aea989d1100ff8 -> 35b0964954afb88a4ac0e64ebe777f194946bf38 ([`pr/nofake.12`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.12) -> [`pr/nofake.13`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.12..pr/nofake.13)) incorporating a different fix for the bug reported https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1505114232 triggered by test code in https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. Instead of updating CheckBlockIndex asserts to cope with ReceivedBlockTransactions temporarily setting the snapshot block nChainTx to 0 if it is downloaded out of order, it updates ReceivedBlockTransactions not to do that. This avoids problems with the previous fix reported https://github.com/bitcoin/bitcoin/pull/29370#pullrequestreview-1912039200 which made CheckBlockIndex checks more complicated

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1986395287

> Seems fine to me, though I'm not in favor of adding nontrivial asserts like this outside of `CheckBlockIndex` - I'd prefer them to be `Assume` in production code, since the consequences of them breaking at some point could be very serious, whereas the consequence of a wrong (non-zero) `nChainTx` value would be mostly just that progress estimation logging would be broken.

Updated 35b0964954afb88a4ac0e64ebe777f194946bf38 -> 1e227bc2139bffe6fd3e3f25d0594c9443836420 ([`pr/nofake.13`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.13) -> [`pr/nofake.14`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.13..pr/nofake.14)) replacing assert with Assume

Sjors commented on src/validation.cpp:
```diff
@@ -5579,21 +5576,11 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
     // Don't make any modifications to the genesis block.
-    // This is especially important because we don't want to erroneously
-    // apply BLOCK_ASSUMED_VALID to genesis, which would happen if we didn't skip
-    // it here (since it apparently isn't BLOCK_VALID_SCRIPTS).
```
One useful bit of this comment might be worth preserving:

```h
// Note that the genesis block isn't BLOCK_VALID_SCRIPTS.
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520015210

> One useful bit of this comment might be worth preserving:

Thanks, added this back

Sjors commented on test/functional/feature_assumeutxo.py:
```diff
@@ -247,6 +254,15 @@ def check_tx_counts(final: bool) -> None:
 
         assert_equal(n1.getblockchaininfo()["blocks"], SNAPSHOT_BASE_HEIGHT)
 
+        self.log.info("Submit a stale block that forked off the chain before the snapshot")
```
85233b203231f39c4436017860248ff9ad08ad42: This is the test mentioned in the PR description? Maybe introduce it in a separate commit so it's a bit easier to test the bug fix.

When I undo the changes in `chain.h` and `validation.cpp` the test fails earlier (line 243), so it tricky to test this part now.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520020544

> [85233b2](https://github.com/bitcoin/bitcoin/commit/85233b203231f39c4436017860248ff9ad08ad42): This is the test mentioned in the PR description? Maybe introduce it in a separate commit so it's a bit easier to test the bug fix.

Good idea, I moved the tests for the two crashes into separate commits.

Sjors replied:
Nice, I was able to produce the crashes - by reverting the main commit and then keeping _one_ of the two crash test commits. Didn't look at the crash log itself.

Sjors reviewed (commented):
I'm still reviewing, but I like how the last commit 065952806234e5a0a6b87df5a5067d0f4062d25e makes it clear that quite a bit of complexity goes away.

Sjors commented on src/validation.cpp:
```diff
@@ -5115,6 +5115,11 @@ double GuessVerificationProgress(const ChainTxData& data, const CBlockIndex *pin
 
+    if (!Assume(pindex->nChainTx > 0)) {
+        LogWarning("Internal bug detected: block %d has unset nChainTx (%s %s). Please report this issue here: %s\n",
+                   pindex->nHeight, PACKAGE_NAME, FormatFullVersion(), PACKAGE_BUGREPORT);
```
c015050cb2cf35303c05c7eb86d3546403e330eb: maybe also `return 0.0;`?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520201099

> [c015050](https://github.com/bitcoin/bitcoin/commit/c015050cb2cf35303c05c7eb86d3546403e330eb): maybe also `return 0.0;`?

Good idea, added

ryanofsky reviewed (commented):
Updated 065952806234e5a0a6b87df5a5067d0f4062d25e -> 69f9f5df4df73d02c4cd184addc36b9fc8f3f507 ([`pr/nofake.15`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.15) -> [`pr/nofake.16`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.15..pr/nofake.16)) adding back a comment and splitting commits as suggested.

Sjors commented on src/validation.cpp:
```diff
@@ -4955,19 +4955,35 @@ void ChainstateManager::CheckBlockIndex()
-                    // even if some data has been pruned.
+                    // If pindex and all its parents downloaded transactions, and
+                    // the transactions were not pruned (pindexFirstMissing is
+                    // null), it is a potential candidate.
```
49426f84b9a15aaae1b3a3c9c5d2ac4ee5855b11: the description matches the code, but I'm a bit confused why an index with a pruned ancestor can't be a candidate. Or is it the case that it _may_ be a candidate, but we just don't have a check for that?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520231273

> [49426f8](https://github.com/bitcoin/bitcoin/commit/49426f84b9a15aaae1b3a3c9c5d2ac4ee5855b11): the description matches the code, but I'm a bit confused why an index with a pruned ancestor can't be a candidate. Or is it the case that it _may_ be a candidate, but we just don't have a check for that?

I think it's the latter. It may be a candidate but we can't have a check ensuring that it is a candidate because if blocks are missing between pindex and the current chain tip, FindMostWorkChain could remove it from the candidate set:

https://github.com/bitcoin/bitcoin/blob/a945f09fa6e0f94cc424da9e06516f9cfa192545/src/validation.cpp#L3073

Probably ideally the check below would be `pindexFirstMissingSinceChainTip == nullptr` instead of `pindexFirstMissing == nullptr`, but it would take more work to be able to write that.

Sjors replied:
That makes sense. So instead of checking if there is a block with missing data between the tip and the new candidate, the code here checks if there's any missing data between genesis and the new candidate. Which is typically the case on a pruned node, which means this check is useless on a pruned node. But that's outside the scope of this PR.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1520231273

> That makes sense. So instead of checking if there is a block with missing data between the tip and the new candidate, the code here checks if there's any missing data between genesis and the new candidate.

Yes, it's pretty confusing so I added some more comments to help clarify.

Sjors commented on src/validation.cpp:
```diff
@@ -4852,13 +4871,30 @@ void ChainstateManager::CheckBlockIndex()
     CBlockIndex* pindexFirstInvalid = nullptr; // Oldest ancestor of pindex which is invalid.
-    CBlockIndex* pindexFirstMissing = nullptr; // Oldest ancestor of pindex which does not have BLOCK_HAVE_DATA.
-    CBlockIndex* pindexFirstNeverProcessed = nullptr; // Oldest ancestor of pindex for which nTx == 0.
+    CBlockIndex* pindexFirstMissing = nullptr; // Oldest ancestor of pindex which does not have BLOCK_HAVE_DATA, since assumeutxo snapshot.
```
Do you mean "since assumeutxo snapshot if used"?

Sjors reviewed (commented):
Mostly happy with 69f9f5df4df73d02c4cd184addc36b9fc8f3f507

`snap_update_firsts` makes me a bit dizzy, mostly because I don't understand what the original code is trying to achieve. I'll try to wrap my ahead around that another time.

_Update 2024-03-12:_ ok, I wrapped my ahead around it. The call sites for `snap_update_firsts` are where we are navigating to the next `CBlockIndex` to check.

There are two sets of `pindexFirst...` variables. The first set starts from genesis + 1 and does not involve assumeutxo. The second set starts from `snapshot`. Blocks are checked depth-first starting from genesis + 1. At some point we _may_ encounter a snapshot block, at which point `snap_update_firsts` swaps the `pindexFirst...` variables. Since we may have a few stale blocks below the snapshot, or even entire fork that replaces the snapshot, `snap_update_firsts` is also called on the way down / sideways.

Sjors reviewed (approved):
ACK 69f9f5df4df73d02c4cd184addc36b9fc8f3f507

I tested loading the signet snapshot with `-checkblockindex` (didn't wait for tip or background sync to finish though).

Note for anyone else testing this: the headers pre-sync phase is agonisingly slow, so you may want to do that bit without `-checkblockindex`. I also ran into #29519 where _all_ my peers were stuck downloading only the background thread (fixed by toggling `setnetworkactive`).

ryanofsky reviewed (commented):
Thanks for the review!

Updated 69f9f5df4df73d02c4cd184addc36b9fc8f3f507 -> ce9388084cb130825ab9faf2ee9e5a6b548a1a07 ([`pr/nofake.16`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.16) -> [`pr/nofake.17`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.16..pr/nofake.17)) adding more comments to clarify things and adding early return in GuessVerificationProgress if Assume check fails
Updated ce9388084cb130825ab9faf2ee9e5a6b548a1a07 -> ae0d1c266898dc1e3c2f72e0d7a7d4d57b164f45 ([`pr/nofake.17`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.17) -> [`pr/nofake.18`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.17..pr/nofake.18)) fixing new comments.

Sjors commented:
re-ACK ae0d1c266898dc1e3c2f72e0d7a7d4d57b164f45

ryanofsky commented:
Updated ae0d1c266898dc1e3c2f72e0d7a7d4d57b164f45 -> 37e54d798a29b4798402db5c4163537046caf536 ([`pr/nofake.18`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.18) -> [`pr/nofake.19`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.18..pr/nofake.19)) trying to improve a few comments.
Updated 37e54d798a29b4798402db5c4163537046caf536 -> 345ac98bb893da7c1f0b573d4da73711b2139391 ([`pr/nofake.19`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.19) -> [`pr/nofake.20`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.19..pr/nofake.20)) adding back nChainTx = 0 fallback in ReceivedBlockTransactions

fjahr reviewed (commented):
Code looks good to me after a latest pass, I will quickly re-review once this has been rebased.

ryanofsky commented on src/test/util/chainstate.h:
```diff
@@ -91,13 +91,16 @@ CreateAndActivateUTXOSnapshot(
-                // and with nTx > 0 (since we aren't setting the pruned flag);
-                // see CheckBlockIndex().
-                pindex->nStatus |= BLOCK_ASSUMED_VALID;
+                // Remove all data and validity flags by just setting
```
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1526681305

> The comment above (`Reset the HAVE_DATA flags...`) could be removed or at least changed to match the changed code here.

The comment above is still accurate and still describes the main thing this is trying to do. Now just other flags are reset as well. Happy to change any of the comments if you have a specific suggestion though.

fjahr replied:
Before the comment could be clearly matched to what was happening below because the HAVE_DATA flags were unset explicitly. Now they are unset implicitly and the new comment states what is going on again more clearly. So this would be confusing as we would have two split comments making the same statement in different words and I would think I might be missing something special about the HAVE_DATA here.

To keep it simple, I would change this:

```
            // Reset the HAVE_DATA flags below the snapshot height, simulating
            // never-having-downloaded them in the first place.
```

to this:

```
            // Simulate never-having-downloaded the blocks below the snapshot height.
```

fjahr replied:
It's a nit, so feel free to keep as is unless you have to retouch.

ryanofsky reviewed (commented):
Rebased 345ac98bb893da7c1f0b573d4da73711b2139391 -> 9d9a7458a2570f7db56ab626b22010591089c312 ([`pr/nofake.20`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.20) -> [`pr/nofake.21`](https://github.com/ryanofsky/bitcoin/commits/pr/nofake.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nofake.20-rebase..pr/nofake.21)) due to conflict with #29478

Sjors commented:
re-ACK 9d9a7458a2570f7db56ab626b22010591089c312

I also tried generating and loading a signet snapshot with this PR combined with #29612.

mzumsande reviewed (commented):
Tested ACK 9d9a7458a2570f7db56ab626b22010591089c312

I reviewed the code and tested this running with `checkblockindex` on signet.

maflcko reviewed (approved):
ACK 9d9a7458a2570f7db56ab626b22010591089c312 🎯

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 9d9a7458a2570f7db56ab626b22010591089c312 🎯
m+XWEuj+TMxTAm2vqwQ2KlIlqiGQf/ADm9CzC8dphRCYKKQo4VacrJ9sy4MIZjiOABr94CzwXsoVSmSPlrElBw==
```

</details>

achow101 commented:
ACK 9d9a7458a2570f7db56ab626b22010591089c312

fjahr commented:
Post merge utACK 9d9a7458a2570f7db56ab626b22010591089c312
